### PR TITLE
Implement performance metrics BehaveTree Node

### PR DESCRIPTION
**Pull request by goudcode**
## Description

I've implemented the performance per tree in the Godot engine profiler.
Time is being showed in nanosecconds and tree will be automatically added and removed to the profiler.

It might be better to disable all the metric code when running in release mode, example: use `if OS.is_debug_build()`
Any ideas are welcome.

## Addressed issues
#89 

## Screenshots
afbeelding

### CODE REVIEW

File: addons/beehave/nodes/beehave_tree.gd [hunk 1]
```diff
@@ -22,6 +22,8 @@ enum {
 
 var actor : Node
 
+var _tree_metric_name : String
```
Comment (bitbrain) [inline addons/beehave/nodes/beehave_tree.gd]: we may want to give this a better name: as it is tracking the `process_time` of this tree, perhaps something like this:
```suggestion
var _process_time_metric_name : String
```
(we are inside the tree already so using "tree" is redundant)

File: addons/beehave/nodes/beehave_tree.gd [hunk 2]
```diff
@@ -93,3 +100,22 @@ func enable() -> void:
 
 func disable() -> void:
 	self.enabled = false
+
+
+func _enter_tree() -> void:
+	# Get the name of the parent node name for metric
+	var parent_name = get_parent().name
+	_tree_metric_name = "beehave/process_time/%s(%s)" % [parent_name, get_instance_id()]
```
Comment (bitbrain) [inline addons/beehave/nodes/beehave_tree.gd]: From a structural perspective, we may want to have multiple metrics as part of the same tree. Therefore, I'd group them differently:
```suggestion
	_tree_metric_name = "beehave/%s-%s/process_time" % [parent_name, get_instance_id()]
```

File: addons/beehave/nodes/beehave_tree.gd [hunk 3]
```diff
@@ -93,3 +100,22 @@ func enable() -> void:
 
 func disable() -> void:
 	self.enabled = false
+
+
+func _enter_tree() -> void:
+	# Get the name of the parent node name for metric
+	var parent_name = get_parent().name
```
Comment (bitbrain) [inline addons/beehave/nodes/beehave_tree.gd]: Using `get_parent()` here is wrong, as there are different ways to associate a node to a tree (see `actor_node_path` export property)
```suggestion
	var parent_name = actor.name
```

**Review by bitbrain — CHANGES_REQUESTED** [review]
Very good start! Apart from a few smaller comments this looks generally good.

Could we also add additional metrics like described on #89?

| Metric Name | Description |
|---|---|
| `beehave/node_count` | The number of total beehave nodes in the scene |
| `beehave/active_count` | Number of active behaviour trees in the scene |

Also reminder, that there is no need to check for debug mode, as Godot will only query the monitors inside the editor anyways and ensures not to call these in production mode.

File: addons/beehave/nodes/beehave_tree.gd [hunk 4]
```diff
@@ -93,3 +100,22 @@ func enable() -> void:
...
+	Performance.add_custom_monitor(_tree_metric_name, _get_tree_metric)
+
+
+func _exit_tree() -> void:
+	# Remove tree metric from the engine
+	Performance.remove_custom_monitor(_tree_metric_name)
+
+
+# Called by the engine to profile this tree
+func _get_tree_metric() -> float:
```
Comment (bitbrain) [inline addons/beehave/nodes/beehave_tree.gd]: Oh, this method also should be renamed to something like this:
```suggestion
func _get_process_time_metric_value() -> float:
```

**Comment by goudcode** [comment]
The metric is now registered in the ready() function instead of the _enter_tree().
The reason for this is that the actor was not known in that function.

**Review by bitbrain — APPROVED** [review]
Thank you for your contribution!

**Pull request merged**